### ISSUE TITLE
Registration not really removed, even after 'remove last registration' 

### ISSUE THREAD

**Issue by NicoKiaru** [BIOP/ijp-imagetoatlas#155]
Reported by @bmi-lsym:

> I probably found a bug in the ABBA save state functionality.
>
> It seems reproducible, but I did not test all possible combinations 
> yet because DeepSlice website stopped working today.
>
>
> Description of the problem:
>
> The whole brain was aligned using DeepSlice, then elastix.
>
> Some sections at the end of the brain are difficult to align 
> automatically if they lost their cortical parts. Therefore, I usually 
> remove all the previous alignments for those sections (via menu item) 
> and then do manually a BigWarp transform from scratch. Then I "edit 
> the last transform" by adding more alignment points. Export of 
> annotations to QuPath works well after that, no problem, and all looks 
> fine.
>
> However, when I save state after this manual alignment, and then 
> reload it in the newly started ABBA session, the image looks very much 
> distorted.
>
> I think this could be due to incomplete removal of the DeepSlice 
> alignment history (which looked like deleted before I did manual 
> BigWarp), because the green circle of the first transform is visible 
> for that slice. Then there is a green square showing the BigWarp was 
> done. While I expected only the square to be there.
>
>
> As a control, I made a clean new small dataset for ABBA, and did 
> BigWarp for one section, followed by "edit last registration" to add 
> more points. As a result, after saving and re-loading everything looks 
> ok. The difference between this condition and the one above - 
> DeepSlice was done for the whole brain for the case above, and does 
> not look like it was really deleted after saving/reloading.
>
>
> Do you think it is indeed a bug, worth fixing?
>

Just to follow it up, I could fix the project with which I had troubles,
by manually editing the main json file to remove DeepSlice registration
step.
It was still hanging around there.
When reloaded, the green circle disappeared, and the slices looked good
again as I left them after the manual registration step.

Conclusion, seems that deletion of DeepSlice registration from the menu
"Remove the last registration" does not get saved, that seems to be the bug.

**Comment by NicoKiaru** [comment]
Thanks for the detailed report @bmi-lsym!

**Comment by NicoKiaru** [comment]
Most probably fixed in https://github.com/BIOP/ijp-imagetoatlas/commit/d535ec580854d1e67462307db6d978bd6edbbfcb

**Comment by bmi-lsym** [comment]
Yes, thanks a lot, @NicoKiaru, the problem is indeed fixed!

**Comment by NicoKiaru** [comment]
Whoa, are you using the current github version ? Because it's not released to Fiji yet

**Comment by bmi-lsym** [comment]
No, but it started working bug-free soon after the date of issue was brought up, I did just regular Fiji updates with BIOPTP update site... I was convinced that you did it then and updated the Fiji distribution.
If not, it is weird indeed...

**Comment by NicoKiaru** [comment]
To anyone reading here: it's a rare bug. I couldn't find a reliable way to trigger it, thus it's hard to fix. As a safety measure, please save your project often and in different files (v0, v1, etc.). Each state file is cheap in memory and fast to save

**Comment by GuillaumeLeGoc** [comment]
Hey Nico,
Just wanted to warn that this bug happened again to me. I loaded my QuPath project, tried DeepSlice (local), move slices around, applied DeepSlice (local) again. Finally I was not conviced so I removed last registration for all slice, save the file, and carried on in mostly-manual mode : interactive transform and BigWarp (sometime affine elastix replaces interactive transform). I saved after processing a single slice in different files at every slice. I registered 10 slices.
New day, I loaded the last state file and the two DeepSlice registrations weren't removed for any slices. I tried my 10 save files, they all have it. Tried to remove the last registration on slices I didn't manually register, saved again, loaded it back and the DeepSlice registrations are still there.
I played around, make some elastix affine registrations and BigWarp, save the state file, remove the last registration on those, save the state file and load it back -- those registrations are properly removed. Only the DeepSlice registrations are not deleted.
The only way to get rid of them is to manually edit the state.json file in the state file.

You'll find enclosed the affected file (renamed .zip to be able to upload it to Github).
[gn53_full_g11.zip](https://github.com/BIOP/ijp-imagetoatlas/files/14021754/gn53_full_g11.zip)


OS and Dependencies Info
```
OS Windows 10
ImageJ 2.14.0
IJ 1.54f
ABBA 0.8.0
BigWarp 7.0.7
Bdv 10.4.6
Bdv Vistools 1.0.0-beta-32
Bdv Biop Tools 0.8.2
Bdv Playground 0.8.1
Biop Image Loader 0.6.4
Biop Wrappers 0.9.6
Updates sites: ImageJ Fiji Java-8 Bio-Formats FFMPEG ilastik Neuroanatomy PTBIOP 
```

**Comment by GuillaumeLeGoc** [comment]
Hey again !
I think I have good news, I played around with the state.json file and stuff.

I _think_ I might have pinpoint the underlying bug. It is not related to DeepSlice, but to any registrations : 
a registration (tried with elastix affine and DeepSlice but any `"type": "RegisterSliceAction"` entry in the state.json file should behave the same) is not really removed when saving the abba state file **when the slice is moved after the registration**, and then "Remove last registration" is done.
In other words, if there are `"type": "MoveSliceAction"` entries in the state.json file _after_ the registration we want to remove, it won't be.

We thought it was due to DeepSlice in particular because the latter is able to move the slices around, and an user is tempted to adjust the rostro-caudal position after the automatic registration and rostro-caudal positioning. On the other hand, when we begin Elastix registration, it's because we're sure of the positions and won't move the slices afterwards.
